File: code/0167. Two Sum II - Input Array Is Sorted/solution1.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int p1 = 0, p2 = numbers.size() - 1;
        while (p1 < p2) {
            int num1 = numbers[p1], num2 = numbers[p2];
            if (num1 + num2 == target) {
                return {p1 + 1, p2 + 1};
            }
            else if (num1 + num2 < target) {
                p1++;
            }
            else {
                p2--;
            }
        }

        return {-1, -1};
    }
};

File: code/0259. 3Sum Smaller/solution1.java
class Solution {
    public int threeSumSmaller(int[] nums, int target) {
        int n = nums.length, result = 0;
        Arrays.sort(nums);
        
        for (int i = 0; i < n; i++) {
            int num1 = nums[i];
            for (int j = i + 1; j < n - 1; j++) {
                int num2 = nums[j];
                if (num1 + num2 + nums[j + 1] >= target) {
                    break;
                }
                
                int left = j + 1, right = n - 1;
                while (left < right) {
                    int mid = (left + right + 1) / 2;
                    if (num1 + num2 + nums[mid] < target) {
                        left = mid;
                    }
                    else {
                        right = mid - 1;
                    }
                }
                result += left - j;
            }
        }
        
        return result;
    }
}